// textlint/lib/engine/rule-set.js
// LICENSE : MIT
"use strict";
// dictionary dictionary
Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var RuleSet = (function () {
    function RuleSet() {
        _classCallCheck(this, RuleSet);

        /**
         * @typedef {{key: Function}} RulesObject
         */
        /**
         * Defined all rules in this object.
         * @type {RulesObject}
         */
        this.rules = Object.create(null);
    }

    /**
     * has rule at least one > 0
     * @returns {boolean}
     */

    _createClass(RuleSet, [{
        key: "hasRuleAtLeastOne",
        value: function hasRuleAtLeastOne() {
            return this.getAllRuleNames().length > 0;
        }
    }, {
        key: "getAllRuleNames",
        value: function getAllRuleNames() {
            return Object.keys(this.rules);
        }
    }, {
        key: "getRule",
        value: function getRule(ruleKey) {
            return this.rules[ruleKey];
        }

        /**
         * @returns {RulesObject}
         */
    }, {
        key: "getAllRules",
        value: function getAllRules() {
            return this.rules;
        }
    }, {
        key: "isDefinedRule",
        value: function isDefinedRule(ruleKey) {
            return this.getRule(ruleKey) != null;
        }

        /**
         *
         * @param {string} ruleKey
         * @param ruleHandler
         */
    }, {
        key: "defineRule",
        value: function defineRule(ruleKey, ruleHandler) {
            this.rules[ruleKey] = ruleHandler;
        }

        /**
         * reset defined rules
         */
    }, {
        key: "resetRules",
        value: function resetRules() {
            this.rules = Object.create(null);
        }
    }]);

    return RuleSet;
})();

exports["default"] = RuleSet;
module.exports = exports["default"];
