// textlint-rule-no-mix-dearu-desumasu/lib/no-mix-dearu-desumasu.js
// LICENSE : MIT
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports["default"] = noMixDearuDesumasu;

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var _textlintRuleHelper = require("textlint-rule-helper");

var _analyzeDesumasuDearu = require("analyze-desumasu-dearu");

function noMixDearuDesumasu(context) {
    var _ref;

    var Syntax = context.Syntax;
    var RuleError = context.RuleError;
    var report = context.report;
    var getSource = context.getSource;

    var helper = new _textlintRuleHelper.RuleHelper(context);
    var dearuCount = undefined;
    var desumasuCount = undefined;

    var dearuLastHit = undefined;
    var desumasuLastHit = undefined;

    function initialize() {
        dearuCount = 0;
        desumasuCount = 0;

        dearuLastHit = null;
        desumasuLastHit = null;
    }

    return (_ref = {}, _defineProperty(_ref, Syntax.Document, initialize), _defineProperty(_ref, Syntax.Str, function (node) {
        if (helper.isChildNode(node, [Syntax.Link, Syntax.Image, Syntax.BlockQuote, Syntax.Emphasis])) {
            return;
        }
        // 例外) 「です・ます」調の文中の「箇条書き」の部分に「である」調を使う場合
        // http://www.p-press.jp/correct/mailmagazine/mailmagazine24.html
        if (helper.isChildNode(node, [Syntax.ListItem])) {
            return;
        }
        var beforeDearuCount = dearuCount;
        var beforeDesumasuCount = desumasuCount;
        var text = getSource(node);
        var retDearu = (0, _analyzeDesumasuDearu.analyzeDearu)(text, { analyzeConjunction: false });
        var retDesumasu = (0, _analyzeDesumasuDearu.analyzeDesumasu)(text, { analyzeConjunction: false });
        dearuCount += retDearu.length;
        desumasuCount += retDesumasu.length;
        if (beforeDearuCount !== dearuCount) {
            dearuLastHit = {
                node: node,
                matches: retDearu
            };
        }
        if (beforeDesumasuCount !== desumasuCount) {
            desumasuLastHit = {
                node: node,
                matches: retDesumasu
            };
        }
    }), _defineProperty(_ref, Syntax.Document + ":exit", function (node) {
        if (dearuCount === 0 || desumasuCount === 0) {
            // No problem
            return;
        }
        if (dearuCount > desumasuCount) {
            // である優先 => 最後の"ですます"を表示
            var hitNode = desumasuLastHit.matches[0];
            var ruleError = new RuleError("\"である\"調 と \"ですます\"調 が混在\n=> \"" + hitNode.value + "\" がですます調\nTotal:\nである  : " + dearuCount + "\nですます: " + desumasuCount + "\n", {
                line: hitNode.lineNumber - 1,
                column: hitNode.columnIndex
            });
            report(desumasuLastHit.node, ruleError);
        } else if (dearuLastHit.matches) {
            // ですます優先 => 最後の"である"を表示
            var hitNode = dearuLastHit.matches[0];
            var ruleError = new RuleError("\"である\"調 と \"ですます\"調 が混在\n=> \"" + hitNode.value + "\" がである調\nTotal:\nである  : " + dearuCount + "\nですます: " + desumasuCount + "\n", {
                line: hitNode.lineNumber - 1,
                column: hitNode.columnIndex
            });

            report(dearuLastHit.node, ruleError);
        }
    }), _ref);
}

module.exports = exports["default"];
