// textlint/lib/core/rule-context.js
// LICENSE : MIT
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports["default"] = RuleContext;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var _fixerRuleFixerCommaner = require("../fixer/rule-fixer-commaner");

var _fixerRuleFixerCommaner2 = _interopRequireDefault(_fixerRuleFixerCommaner);

var _ruleError = require("./rule-error");

var _ruleError2 = _interopRequireDefault(_ruleError);

var _sharedRuleSeverity = require("../shared/rule-severity");

/**
 * Rule context object is passed to each rule as `context`
 * @param ruleId
 * @param sourceCode
 * @param report
 * @param textLintConfig
 * @param ruleConfig
 * @returns {*}
 * @constructor
 */
var assert = require("assert");

function RuleContext(ruleId, sourceCode, report, textLintConfig, ruleConfig) {
    Object.defineProperty(this, "id", { value: ruleId });
    Object.defineProperty(this, "config", { value: textLintConfig });
    var severity = (0, _sharedRuleSeverity.getSeverity)(ruleConfig);
    /**
     *
     * @param {TxtNode} node
     * @param {RuleError|any} error error is a RuleError instance or any data
     */
    this.report = function (node, error) {
        assert(!(node instanceof _ruleError2["default"]), "should be `report(node, ruleError);`");
        if (error instanceof _ruleError2["default"]) {
            report({ ruleId: ruleId, node: node, severity: severity, error: error });
        } else {
            var level = error.severity || _sharedRuleSeverity.SeverityLevel.error;
            report({ ruleId: ruleId, node: node, severity: level, error: error });
        }
    };
    // Const Values
    Object.defineProperty(this, "Syntax", {
        get: function get() {
            return sourceCode.getSyntax();
        }
    });
    this.getFilePath = sourceCode.getFilePath.bind(sourceCode);
    this.getSource = sourceCode.getSource.bind(sourceCode);
    // CustomError object
    this.RuleError = _ruleError2["default"];
    // fixer
    this.fixer = new _fixerRuleFixerCommaner2["default"]();
}

module.exports = exports["default"];
